jwst/lib/pipe_utils.py

```diff
@@ -76,3 +83,69 @@ def is_irs2(model):
         return True
     else:
         return False
+
+
+def match_nans_and_flags(input_model):
+    """Ensure data, error, variance, and DQ are marked consistently for invalid data.
+
+    Invalid data is assumed to be any pixel set to NaN in any one of the
+    data, error, or variance arrays, or else set to the DO_NOT_USE flag
+    in the DQ array.
+
+    The input model is updated in place with NaNs or DO_NOT_USE flags, as
+    appropriate, at all invalid data locations.
+
+    Parameters
+    ----------
+    input_model : DataModel
+        Input model containing some combination of data, dq, err, var_rnoise,
+        var_poisson, and var_flat extensions. These extensions must all have
+        matching dimensions if present.
+    """
+    # Build up the invalid data flags from each available data extension.
+    is_invalid = None
+    data_shape = None
+    nan_extensions = ['data', 'err', 'var_rnoise', 'var_poisson', 'var_flat']
+    for extension in nan_extensions:
+        if not hasattr(input_model, extension):
+            continue
+        data = getattr(input_model, extension)
+        if is_invalid is None:
+            is_invalid = np.isnan(data)
+            data_shape = data.shape
+        else:
+            if data.shape != data_shape:
+                log.warning(f"Mismatched data shapes; skipping invalid data "
```

Comment:
Are there any cases where we would expect one of these arrays (say, `var_rnoise`) to be the wrong shape but still trust the results of this?  Put another way, isn't any shape mismatch a sign of a different issue earlier in the pipeline flow?  I guess I would expect this to throw an error.

Reply:
This is not expected in normal use cases in the pipeline, but there are a couple unit tests for the resampling step that pass in a mismatched variance (with size 0) and expect the step to finish successfully, with a warning for the missing variance.